File: archer2-GPU-directives/Exercises/pi/Fortran/pi.f90
program pi_fortran
	use omp_lib
	use iso_fortran_env, only : int64
	implicit none

	integer, parameter :: dp = kind(1.0d0)

	integer (kind = int64) :: i, num_steps
	real (kind = dp) :: x, pi, start, end, sum, step

	num_steps = 500000000_int64
	sum = 0.0
	step = 1.0_dp / num_steps
	
	start = omp_get_wtime()
	
	do i = 0, num_steps - 1
		x = (i + 0.5) * step
		sum = sum + 4.0_dp / (1.0_dp + x**2)
	end do

	end = omp_get_wtime()
	
! calculate and output results
	pi = step * sum
	write (*, *) "Pi = ", pi
	write (*, *) "Total Time =", end - start, "(sec)"

end program pi_fortran

File: archer2-GPU-directives/Exercises/pi/C/pi.c
#include <stdio.h>
#include <omp.h>

static long num_steps = 500000000;

int main ()
{
	double x, pi; 
	double sum = 0.0;
	double step = 1.0 / (double)num_steps;

	double start, end;
	start = omp_get_wtime();
        //#pragma omp parallel for default(none) private(x) shared(pi,step) reduction(+:sum) 
	#pragma omp target teams distribute parallel for map(tofrom:sum) map(to:step,num_steps) reduction(+:sum)  
        for (long i = 0; i < num_steps; i++) 
	{
		x = (i + 0.5) * step;
		sum = sum + 4.0 / (1.0 + x*x);
	}

	end = omp_get_wtime();
	
	pi = step * sum;
	printf("Pi = %.14lf\n", pi); 
	printf("Total Time = %f (sec)\n", end - start);
}
